kani-compiler/src/kani_middle/transform/check_uninit/ty_layout.rs

```diff
@@ -332,3 +332,36 @@ fn data_bytes_for_ty(
         FieldsShape::Array { .. } => Ok(vec![]),
     }
 }
+
+/// Returns true if `to_ty` has a smaller or equal size and the same padding bytes as `from_ty` up until
+/// its size.
+pub fn tys_layout_compatible(from_ty: &Ty, to_ty: &Ty) -> bool {
+    // Retrieve layouts to assess compatibility.
+    let from_ty_info = PointeeInfo::from_ty(*from_ty);
+    let to_ty_info = PointeeInfo::from_ty(*to_ty);
+    if let (Ok(from_ty_info), Ok(to_ty_info)) = (from_ty_info, to_ty_info) {
+        let from_ty_layout = match from_ty_info.layout() {
+            PointeeLayout::Sized { layout } => layout,
+            PointeeLayout::Slice { element_layout } => element_layout,
+            PointeeLayout::TraitObject => return false,
+        };
+        let to_ty_layout = match to_ty_info.layout() {
+            PointeeLayout::Sized { layout } => layout,
+            PointeeLayout::Slice { element_layout } => element_layout,
+            PointeeLayout::TraitObject => return false,
+        };
+        // Ensure `to_ty_layout` does not have a larger size.
+        if to_ty_layout.len() <= from_ty_layout.len() {
```

Comment:
What about the slice case? We are completely ignoring the size of the slice.

Reply:
Is it possible to affect that in any way with `transmute`?

Reply:
If you are making an assumption that a scenario is not possible, please add an assert.

Reply:
I meant that I think this is always satisfied, i.e. does not need to be checked. Transmuting a slice is valid as long as the element padding is compatible.